File: backend/app/services/aito_board_rules.py
"""The Aito board's rules: a project's column is derived, not dropped.

A project sits where its quote status and its ticked task steps say it sits.
This module is the authoritative definition of that, and it is pure (no
FastAPI, no SQLAlchemy, no models) so it can be unit-tested exhaustively.

It IS mirrored in TypeScript — frontend/src/utils/aitoBoardRules.ts — because
the board is optimistic: the card has to move the instant a step is ticked,
which means the frontend must predict the column rather than wait to be told
it. The mirror is not maintained by discipline. It is pinned by a generated
contract fixture (backend/tests/aito_rules_fixture.py), so changing anything
here without updating the mirror fails the build. After editing this file run:

    ./venv/bin/python3 scripts/gen_aito_board_rules_fixture.py

and fix the TypeScript until the frontend suite is green again.
"""

from collections.abc import Collection, Iterable
from dataclasses import dataclass
from typing import Any

# Board order, left to right. `waiting`, `scan` and `done` were added and
# `pickup` removed on 2026-07-29; see the migration in core/database.py.
COLUMN_ORDER: tuple[str, ...] = ("devis", "waiting", "scan", "model", "print", "finish", "done")

# Zoho statuses meaning the quote has left the shop: the answer is now the
# client's to give, not ours to write. `viewed` only says they opened it and
# `expired` says they never answered — both are still waiting on them.
AWAY_STATUSES: frozenset[str] = frozenset({"sent", "viewed", "expired"})

# The four services a task can carry, in the canonical order the rest of the
# Aito code emits them. This is the canonical order — api/routes/aito.py
# imports SERVICES from here rather than defining its own.
SERVICES: tuple[str, ...] = ("scan", "modelisation", "impression", "usinage")

# Which services each work stage covers, in board order. Printing and machining
# share one column while remaining two separate steps on a task: the column is
# left only once BOTH are ticked everywhere they appear.
STAGES: tuple[tuple[str, tuple[str, ...]], ...] = (
    ("scan", ("scan",)),
    ("model", ("modelisation",)),
    ("print", ("impression", "usinage")),
)


@dataclass(frozen=True)
class TaskSteps:
    """One task's steps, as the card draws them.

    ``services`` is what the task carries, ``done`` the ticked subset, both in
    ``SERVICES`` order. Deliberately the same ``(services, done)`` pair the
    frontend's ``ServiceBadges`` takes, so the API, the board card and a
    collapsed task row all describe a task's steps the same way.
    """

    services: tuple[str, ...] = ()
    done: tuple[str, ...] = ()


@dataclass(frozen=True)
class TaskSummary:
    """Everything a project's tasks say about it, in one value.

    ``services`` and ``pending`` are in ``SERVICES`` order so the card's badge
    row is stable across refetches regardless of the order tasks were created
    in. ``total`` is the definition mirrored by ``taskTotal`` in
    frontend/src/utils/taskDraft.ts.

    ``steps_total``/``steps_done`` count (task, service) pairs, not services:
    two tasks each carrying a scan are two steps, where ``services`` would
    report ``('scan',)`` once. They are what the board card's progress bar
    reads. A service priced at 0 is a real step; a service priced ``None`` is
    absent from the job and is not counted at all, done flag or no.

    This whole dataclass is mirrored by ``summariseTasks`` in
    frontend/src/utils/aitoBoardRules.ts and pinned by the contract fixture —
    see backend/tests/aito_rules_fixture.py. Changing it here without
    regenerating that fixture fails the build, by design.
    """

    count: int = 0
    total: float = 0.0
    services: tuple[str, ...] = ()
    pending: tuple[str, ...] = ()
    steps_total: int = 0
    steps_done: int = 0
    # One entry per task, in the order the caller handed them over — the card
    # draws a row per entry and the detail panel lists them in that same order.
    # This is what makes the card's pill grid possible without shipping every
    # task row on GET /aito/.
    steps_by_task: tuple[TaskSteps, ...] = ()


def summarise(tasks: Iterable[Any]) -> TaskSummary:
    """Count, total, enabled services and pending services in one pass.

    Duck-typed over anything exposing ``<service>_cost`` and ``<service>_done``
    — an ``AitoTask`` row in practice — so this module never imports a model.

    A cost of ``None`` means the service is absent from the job and is skipped
    entirely; ``0`` means it is quoted free, which is a real step that must
    show its badge and hold its column.
    """
    rows = list(tasks)
    total = 0.0
    enabled: set[str] = set()
    unticked: set[str] = set()
    steps_total = 0
    steps_done = 0
    by_task: list[TaskSteps] = []
    for task in rows:
        task_services: list[str] = []
        task_done: list[str] = []
        for service in SERVICES:
            cost = getattr(task, f"{service}_cost")
            if cost is None:
                continue
            enabled.add(service)
            task_services.append(service)
            total += cost
            steps_total += 1
            if getattr(task, f"{service}_done"):
                steps_done += 1
                task_done.append(service)
            else:
                unticked.add(service)
        by_task.append(TaskSteps(services=tuple(task_services), done=tuple(task_done)))
    return TaskSummary(
        count=len(rows),
        total=total,
        services=tuple(service for service in SERVICES if service in enabled),
        pending=tuple(service for service in SERVICES if service in unticked),
        steps_total=steps_total,
        steps_done=steps_done,
        steps_by_task=tuple(by_task),
    )


def evaluate(quote_status: str | None, stored_column: str, pending: Collection[str]) -> tuple[str, str | None]:
    """The whole rule set: ``(column, move_lock)``.

    ``move_lock`` names why the card cannot be dragged between columns, and is
    ``None`` only when it can (Finish <-> Done). It is what the card renders its
    lock badge and tooltip from, and what decides which droppables a drag
    enables — mirrored, not independently re-derived, by the frontend's own
    ``evaluate`` in frontend/src/utils/aitoBoardRules.ts (see the module
    docstring above for how that mirror is pinned).

    Rule order matters twice. Waiting outranks the steps, so ticking a step on
    a card that is out with the client moves nothing — the work is not
    authorised yet. And the stage search runs before the nothing-left-to-do
    fallback, which is what evicts a card from Done the moment any step is
    re-opened; swapped, un-ticking would leave it parked in Done forever.
    """
    if quote_status == "declined":
        return "done", "declined"
    if quote_status in AWAY_STATUSES:
        return "waiting", "waiting"
    if quote_status != "accepted":
        # NULL included: a hand-made card with no Zoho quote waits for Accept
        # exactly like a draft does. Acceptance is the single gate.
        return "devis", "quote"

    pending_set = set(pending)
    for stage, services in STAGES:
        if pending_set.intersection(services):
            return stage, "steps"

    # Nothing left to do. This is the ONLY place the stored column is believed,
    # and only between Finish and Done — which is what makes that one manual
    # drag possible inside an otherwise fully derived model.
    return ("done" if stored_column == "done" else "finish"), None

File: frontend/src/utils/aitoBoardRules.ts
import type { AitoColumnId } from '../api/client';

/** The Aito board's rules, mirrored from
 *  backend/app/services/aito_board_rules.py.
 *
 *  A mirror exists because the board is optimistic: a card must move the
 *  instant a step is ticked or a quote is accepted, which means predicting the
 *  column locally rather than waiting to be told it.
 *
 *  This mirror is NOT maintained by discipline. It is pinned by a generated
 *  contract fixture — see backend/tests/aito_rules_fixture.py. Change the
 *  Python and the backend test fails until the fixture is regenerated;
 *  regenerate it and this file's test fails until it is brought back in line.
 *
 *  Imports nothing local but a type. `services.ts` and `taskDraft.ts` both
 *  import FROM here; importing either of them back would close a cycle. */

export type MoveLock = 'quote' | 'waiting' | 'declined' | 'steps' | null;

/** Board order, left to right. */
export const COLUMN_ORDER: readonly AitoColumnId[] = [
  'devis',
  'waiting',
  'scan',
  'model',
  'print',
  'finish',
  'done',
];

/** Canonical service order — every derived list is emitted in it, so a badge
 *  row is stable across refetches regardless of task creation order.
 *
 *  This is the single declaration of which services exist. `ServiceId` below
 *  is derived FROM this array (`as const` + indexed access) rather than
 *  hand-typed alongside it, and deliberately so: a plain `Record<string, ...>`
 *  keyed by service, or a union typed by hand next to this array, would let a
 *  fifth service compile into `ServiceId` without ever landing here — and
 *  every lookup keyed by the union (`COST_KEYS`, `TaskLike.done`) would then
 *  miss it silently at runtime, `undefined` with no compile error to catch
 *  it. Making `SERVICES` authoritative closes that: a service can only be
 *  added by editing this array, which immediately fails compilation of every
 *  `Record<ServiceId, ...>` until it is filled in too. `STAGES` below can't be
 *  pinned this way — TypeScript has no way to express "every union member
 *  appears in this nested array" — so it is covered by a test instead
 *  (aitoBoardRules.test.ts: "every service is staged exactly once"). */
export const SERVICES = ['scan', 'modelisation', 'impression', 'usinage'] as const;

export type ServiceId = (typeof SERVICES)[number];

/** Statuses meaning the quote has left the shop: the answer is the client's to
 *  give, not ours to write. `viewed` only says they opened it and `expired`
 *  says they never answered — both are still waiting on them. */
export const AWAY_STATUSES: ReadonlySet<string> = new Set(['sent', 'viewed', 'expired']);

/** Which services each work stage covers, in board order. Printing and
 *  machining share one column while remaining two separate steps on a task:
 *  the column is left only once BOTH are ticked everywhere they appear.
 *
 *  Exported only so the contract test can assert every member of `SERVICES`
 *  appears in exactly one stage — unlike `ServiceId`, that completeness can't
 *  be pinned by the type system, since nothing stops a tuple entry from being
 *  duplicated or omitted while still type-checking as `readonly ServiceId[]`. */
export const STAGES: readonly (readonly [AitoColumnId, readonly ServiceId[]])[] = [
  ['scan', ['scan']],
  ['model', ['modelisation']],
  ['print', ['impression', 'usinage']],
];

/** The minimum a task must expose for these rules to read it. Structural, not
 *  `TaskDraft`, for the same reason the Python is duck-typed: it keeps this
 *  module free of any dependency that could cycle back into it. */
export interface TaskLike {
  scanCost: number | null;
  modelisationCost: number | null;
  impressionCost: number | null;
  usinageCost: number | null;
  done: Record<ServiceId, boolean>;
}

const COST_KEYS: Record<ServiceId, keyof TaskLike> = {
  scan: 'scanCost',
  modelisation: 'modelisationCost',
  impression: 'impressionCost',
  usinage: 'usinageCost',
};

/** One service's cost, or null when the service is absent from the job.
 *  `0` is a real cost — a step quoted free. */
export function taskCost(task: TaskLike, service: ServiceId): number | null {
  return task[COST_KEYS[service]] as number | null;
}

/** The whole rule set: `[column, moveLock]`.
 *
 *  `moveLock` names why the card is still pinned to a derived column, and is
 *  null only once the rules let go of it — which happens exactly between
 *  Finish and Done. That release is what gates the card's mark-done and
 *  restore buttons; nothing about it is a drag any more.
 *
 *  Rule ORDER matters twice, and is the part a data-driven version could not
 *  express. Waiting outranks the steps, so ticking a step on a card that is
 *  out with the client moves nothing — the work is not authorised yet. And the
 *  stage search runs before the nothing-left-to-do fallback, which is what
 *  evicts a card from Done the moment any step is re-opened; swapped,
 *  un-ticking would leave it parked in Done forever. */
export function evaluate(
  quoteStatus: string | null,
  storedColumn: AitoColumnId,
  pending: readonly string[],
): [AitoColumnId, MoveLock] {
  if (quoteStatus === 'declined') return ['done', 'declined'];
  if (quoteStatus !== null && AWAY_STATUSES.has(quoteStatus)) return ['waiting', 'waiting'];
  if (quoteStatus !== 'accepted') {
    // null included: a hand-made card with no Zoho quote waits for Accept
    // exactly like a draft does. Acceptance is the single gate.
    return ['devis', 'quote'];
  }

  const pendingSet = new Set(pending);
  for (const [stage, services] of STAGES) {
    if (services.some((service) => pendingSet.has(service))) return [stage, 'steps'];
  }

  // Nothing left to do. This is the ONLY place the stored column is believed,
  // and only between Finish and Done — which is what makes that one manual
  // transition (the card's mark-done / restore buttons) possible inside an
  // otherwise fully derived model.
  return [storedColumn === 'done' ? 'done' : 'finish', null];
}

/** One task's steps, as the card draws them. Mirrors `TaskSteps` in
 *  backend/app/services/aito_board_rules.py. `services` is what the task
 *  carries and `done` the ticked subset, both in `SERVICES` order. */
export interface TaskSteps {
  services: ServiceId[];
  done: ServiceId[];
}

export interface TaskSummary {
  count: number;
  total: number;
  services: ServiceId[];
  pending: ServiceId[];
  stepsTotal: number;
  stepsDone: number;
  /** One entry per task, in the order given — the card draws a row each. */
  stepsByTask: TaskSteps[];
}

/** Everything a project's tasks say about it, in one pass.
 *
 *  A cost of null means the service is absent from the job and is skipped
 *  entirely; 0 means it is quoted free, which is a real step that must show
 *  its badge, hold its column and count toward the progress bar.
 *
 *  `stepsTotal`/`stepsDone` count (task, service) PAIRS, not services: two
 *  tasks each carrying a scan are two steps, where `services` reports 'scan'
 *  once. */
export function summariseTasks(tasks: readonly TaskLike[]): TaskSummary {
  let total = 0;
  let stepsTotal = 0;
  let stepsDone = 0;
  const enabled = new Set<ServiceId>();
  const unticked = new Set<ServiceId>();
  const stepsByTask: TaskSteps[] = [];

  for (const task of tasks) {
    const taskServices: ServiceId[] = [];
    const taskDone: ServiceId[] = [];
    for (const service of SERVICES) {
      const cost = taskCost(task, service);
      if (cost === null) continue;
      enabled.add(service);
      taskServices.push(service);
      total += cost;
      stepsTotal += 1;
      if (task.done[service]) {
        stepsDone += 1;
        taskDone.push(service);
      } else {
        unticked.add(service);
      }
    }
    stepsByTask.push({ services: taskServices, done: taskDone });
  }

  return {
    count: tasks.length,
    total,
    services: SERVICES.filter((service) => enabled.has(service)),
    pending: SERVICES.filter((service) => unticked.has(service)),
    stepsTotal,
    stepsDone,
    stepsByTask,
  };
}
